cpp/src/arrow/util/align_util_test.cc

```diff
@@ -59,6 +63,18 @@ void CheckBitmapWordAlign(const uint8_t* data, int64_t bit_offset, int64_t lengt
   }
 }
 
+arrow::Result<std::shared_ptr<Array>> CreateUnalignedArray(const Array& array) {
+  BufferVector sliced_buffers(array.data()->buffers.size(), nullptr);
+  for (std::size_t i = 0; i < array.data()->buffers.size(); ++i) {
+    if (array.data()->buffers[i]) {
+      sliced_buffers[i] = SliceBuffer(array.data()->buffers[i], 1, 2);
+    }
+  }
+  auto sliced_array_data =
+      ArrayData::Make(array.type(), array.length(), std::move(sliced_buffers));
```

Comment:
```suggestion
  auto sliced_array_data =
      ArrayData::Make(array.type(), /*length=*/2, std::move(sliced_buffers));
```

This works today, but only because you never call `Validate` or any kind of compute function on the returned array.  This could be a time bomb for a future maintainer.  Better to make sure the array is properly formed which means the length of the array must match the length of its buffers.